pvlib/mlfm.py

```diff
@@ -0,0 +1,803 @@
+
+    See also
+    --------
+    mlfm_meas_to_norm
+
+    References
+    ----------
+    .. [1] Steve Ransome (SRCL) and Juergen Sutterlueti (Gantner Instruments)
+       "Quantifying Long Term PV Performance and Degradation under Real Outdoor
+       and IEC 61853 Test Conditions Using High Quality Module IV Measurements"
+       36th EU PVSEC, Marseille, France. September 2019
+    '''
+
+    # create an empty DataFrame to put stack results
+    dstack = pd.DataFrame()
+
+    # create a gap to differentiate i and v losses : gap width~0.01
+    gap = 0.01
+
+    inv_ff = 1 / fill_factor
+
+    if all(c in dnorm.columns for c in ['v_ff', 'r_oc', 'i_ff', 'r_sc']):
+
+        # include effects of series and shunt resistances in stacked losses
+        # find factor to transform multiplicative to subtractive losses
+        # correction factor to scale losses to keep 1/ff --> pr_dc
+
+        # product
+        prod = inv_ff * (
+            dnorm['i_sc'] * dnorm['r_sc'] * dnorm['i_ff'] *
+            dnorm['v_ff'] * dnorm['r_oc'] * dnorm['v_oc']
+        )
+
+        # total
+        tot = inv_ff + (
+            dnorm['i_sc'] + dnorm['r_sc'] + dnorm['i_ff'] +
+            dnorm['v_ff'] + dnorm['r_oc'] + dnorm['v_oc'] - 6
+        )
+
+        # correction factor
+        corr = (inv_ff - prod) / (inv_ff - tot)
+
+        # put mlfm values in a stack from pr_dc (bottom) to 1/ff_ref (top)
+        # accounting for series and shunt resistance losses
+        dstack['pr_dc'] = +dnorm['pr_dc']  # initialise
+        dstack['i_sc'] = -(dnorm['i_sc'] - 1) * corr
+        dstack['r_sc'] = -(dnorm['r_sc'] - 1) * corr
+        dstack['i_ff'] = -(dnorm['i_ff'] - 1) * corr - gap/2
+        dstack['i_v'] = gap
+        dstack['v_ff'] = -(dnorm['v_ff'] - 1) * corr - gap/2
+        dstack['r_oc'] = -(dnorm['r_oc'] - 1) * corr
+        dstack['v_oc'] = -(dnorm['v_oc'] - 1) * corr
+        dstack['temp_module_corr'] = (
+            -(dnorm['v_oc'] - dnorm['v_oc_temp_corr']) * corr)
+
+        return dstack
+
+    # subtractive losses without series and shunt resistance effects
+    # find factor to transform multiplicative to subtractive losses
+    # correction factor to scale losses to keep 1/ff --> pr_dc
+
+    prod = inv_ff * (
+        dnorm['i_sc'] * dnorm['i_mp'] *
+        dnorm['v_mp'] * dnorm['v_oc']
+    )
+
+    tot = inv_ff + (
+        dnorm['i_sc'] + dnorm['i_mp'] +
+        dnorm['v_mp'] + dnorm['v_oc'] - 4
+    )
+
+    corr = (inv_ff - prod) / (inv_ff - tot)
+
+    # put mlfm values in a stack from pr_dc (bottom) to 1/ff_ref (top)
+    dstack['pr_dc'] = + dnorm['pr_dc']  # initialise
+    dstack['i_sc'] = -(dnorm['i_sc'] - 1) * corr
+    dstack['i_mp'] = -(dnorm['i_mp'] - 1) * corr - gap/2
+    dstack['i_v'] = gap
+    dstack['v_mp'] = -(dnorm['v_mp'] - 1) * corr - gap/2
+    dstack['v_oc'] = -(dnorm['v_oc'] - 1) * corr
+
+    dstack['temp_module_corr'] = (
+        - (dnorm['v_oc'] - dnorm['v_oc_temp_corr']) * corr)
```

Comment:
@steve-ransome can you confirm that these statements are all correct? When I made a new test from the 6-element data, replacing `v_ff` and `r_oc` by `v_mp = v_ff * r_oc` and `i_ff` and `r_sc` by `i_mp = i_ff * r_sc`, the stacked `i_sc` wasn't very close to the value computed from the 6-element test. I had to make a new fixture `stack4` and just copied output from this code block.

Reply:
Thanks Cliff. The norm data is multiplicative so should work with multiplication as I wrote,
The stack data is subtractive (to get the stack plot to show each loss term separately)

norm_v_mp = norm_v_ff * norm_r_oc and norm_i_mp = norm_i_ff * norm_r_sc

stack_v_mp = stack_v_ff + stack_r_oc and stack_i_mp = stack_i_ff + stack_r_sc

Does that work?

Reply:
Not exactly the same values. So I think we need to create new, known output for the 4-variable case. Because I did that using the code referenced above, if you confirm that code is correct, then we'll trust the values in the new fixture. 

```
stack6['v_ff'] + stack6['r_oc']
Out[87]: 
0    0.217098
dtype: float64

stack4['v_mp']
Out[88]: 
0    0.214483
Name: v_mp, dtype: float64
```

Reply:
Thanks Cliff, I checked it and your figures are correct. 

While line 1 is true
1) norm_v_mp = norm_v_ff * norm_r_oc , norm_i_mp = norm_i_ff * norm_r_sc
Line 2 is only approximate (to about 0.2%)
2) stack_v_mp ~ stack_v_ff + stack_r_oc , stack_i_mp ~ stack_i_ff + stack_r_sc

We are taking accurate multiplicative losses and the putting them onto a linear graph so there may be some small distortion but it is small. 

It's the overall relative losses (e.g. r_oc vs. r_sc) and any changes with time that matter. The sum of all the stack losses and the product of all multiplicative losses are the same from stack_6 and stack_4 as that's the way they drop from 1/ff to pr_dc and the stack graphs are very useful.

Therefore we need to check the stack_4 figures as you have done.

Reply:
Cliff, as discussed at the PVPMC the multiplicative LFM losses are all correct (the product of them takes us from the initial 1/ff to pr_dc) but in translating them to subtractive losses where the sum of them takes us from 1/ff to pr_dc correctly, but there's a tiny bit of non-linearity meaning the lfm4 and lfm6 figures won't be exactly the same, they'll be close and as the stack loss is just a visual interpretation of losses should be good enough to identify losses and degradation as in previous publications. If we want to get them exactly the same (not necessary but nice) then we could do stack plots of 1-loss on a log  y scale which shows it well with the thickness of the bar indicating the relative loss but we then might have to rethink possible negative losses such as isc gain due to spectral effects and the log graph would need to go from only 0.5 to 1.5 or so. What's the next stage on this as isn't the other code working OK?